libopenage/util/profiler.cpp

```diff
@@ -0,0 +1,182 @@
+
+void Profiler::unregister_all() {
+	std::vector<std::string> registered_components = this->registered_components();
+
+	for (auto com : registered_components) {
+		this->unregister_component(com);
+	}
+}
+
+std::vector<std::string> Profiler::registered_components() {
+	std::vector<std::string> registered_components;
+	for (auto it = this->components.begin(); it != this->components.end(); ++it) {
+		registered_components.push_back(it->first);
+	}
+
+	return registered_components;
+}
+
+void Profiler::start_measure(std::string com, color component_color) {
+	if (not this->registered(com)) {
+		this->register_component(com, component_color);
+	}
+
+	this->components[com].start = std::chrono::high_resolution_clock::now();
+}
+
+void Profiler::end_measure(std::string com) {
+	if (this->registered(com)) {
+		std::chrono::high_resolution_clock::time_point end = std::chrono::high_resolution_clock::now();
+		this->components[com].duration = end - this->components[com].start;
+	}
+}
+
+void Profiler::draw_component_performance(std::string com) {
+	color rgb = this->components[com].drawing_color;
+	glColor4f(rgb.r, rgb.g, rgb.b, 1.0);
+
+	glLineWidth(1.0);
+	glBegin(GL_LINE_STRIP);
+	float x_offset = 0.0;
+	float offset_factor = (float)PROFILER_CANVAS_WIDTH / (float)MAX_DURATION_HISTORY;
+	float percentage_factor = (float)PROFILER_CANVAS_HEIGHT / 100.0;
+
+	for (auto i = this->insert_pos; keep_in_duration_bound(i) != keep_in_duration_bound(this->insert_pos-1); ++i) {
+		i = keep_in_duration_bound(i);
+
+		auto percentage = this->components[com].history.at(i);
+		glVertex3f(PROFILER_CANVAS_POSITION_X + x_offset, PROFILER_CANVAS_POSITION_Y + percentage * percentage_factor, 0.0);
+		x_offset += offset_factor;
+	}
+	glEnd();
+
+	// reset color
+	glColor4f(1.0, 1.0, 1.0, 1.0);
+}
+
+void Profiler::show(bool debug_mode) {
+	if (debug_mode) {
+		this->show();
+	}
+}
+
+void Profiler::show() {
+
+	this->draw_canvas();
+	this->draw_legend();
+
+	for (auto com : this->components) {
+		this->draw_component_performance(com.first);
+	}
+}
+
+bool Profiler::registered(std::string com) const {
+	return this->components.find(com) != this->components.end();
+}
+
+unsigned Profiler::size() const {
+	return this->components.size();
+}
+
+void Profiler::start_frame_measure() {
+	this->frame_start = std::chrono::high_resolution_clock::now();
+}
+
+void Profiler::end_frame_measure() {
+	auto frame_end = std::chrono::high_resolution_clock::now();
+	this->frame_duration = frame_end - this->frame_start;
+
+	for (auto com : this->registered_components()) {
+		double percentage = this->duration_to_percentage(this->components[com].duration);
+		this->append_to_history(com, percentage);
+	}
+
+	this->insert_pos++;
+}
+
+void Profiler::draw_canvas() {
+	glColor4f(0.2, 0.2, 0.2, PROFILER_CANVAS_ALPHA);
+	glRecti(PROFILER_CANVAS_POSITION_X,
+			PROFILER_CANVAS_POSITION_Y,
+			PROFILER_CANVAS_POSITION_X + PROFILER_CANVAS_WIDTH,
+			PROFILER_CANVAS_POSITION_Y + PROFILER_CANVAS_HEIGHT);
```

Comment:
indent:
tabs to reach position of `glRecti`, spaces to move `PROFILER...` after the `(`